docs/tls.md

```diff
@@ -11,21 +20,57 @@ tls_letsencrypt_cache_dir: ".cache"
+For `TLS-ALPN-01`, headscale listens on the ip:port combination defined in `listen_addr`. Let's Encrypt will _only_ connect to port 443 for the validation callback, so if `listen_addr` is not set to port 443, something else (e.g. a firewall rule) will be required to forward the traffic from port 443 to the ip:port combination specified in `listen_addr`.
 
-## Bring your own certificate
+### Technical description
 
-headscale can also be configured to expose its web service via TLS. To configure the certificate and key file manually, set the `tls_cert_path` and `tls_cert_path` configuration parameters. If the path is relative, it will be interpreted as relative to the directory the configuration file was read from.
+Headscale uses [autocert](https://pkg.go.dev/golang.org/x/crypto/acme/autocert), a Golang library providing [ACME protocol](https://en.wikipedia.org/wiki/Automatic_Certificate_Management_Environment) verification, to facilitate certificate renewals via [Let's Encrypt](https://letsencrypt.org/about/). Certificates will be renewed automatically, and the following can be expected:
 
-```yaml
-tls_cert_path: ""
-tls_key_path: ""
+* Certificates provided from Let's Encrypt have a validity of 3 months from date issued.
+* Renewals are only attempted by headscale when 30 days or less remains until certificate expiry.
+* Renewal attempts by autocert are triggered at a random interval of 30-60 minutes.
+* No log output is generated for failed or successful renewals.
+
+#### Checking certificate expiry
+
+If you want to validate that certificate renewal completed successfully, this can be done either manually, or through external monitoring software. Two examples of doing this manually:
+
+1. Open the URL for your Headscale server in your browser of choice, and manually inspecting the expiry date of the certificate you receive.
+1. Or, check remotely from CLI using `openssl`:
+
+```bash
+$ openssl s_client -servername [hostname] -connect [hostname]:443 | openssl x509 -noout -dates
+(...)
+notBefore=Feb  8 09:48:26 2024 GMT
+notAfter=May  8 09:48:25 2024 GMT
 ```
+
+#### Log output from the autocert library
+
+As these log lines are from the autocert library, they are not strictly generated by headscale itself.
+
+```plaintext
+acme/autocert: missing server name
+```
+
+Likely caused by an incoming connection that does not specify a hostname, for example a `curl` request directly against the IP of the server, or an unexpected hostname.
+
+```plaintext
+acme/autocert: host "[foo]" not configured in HostWhitelist
+```
+
+Similarly to the above, this likely indicates an invalid incoming request for an incorrect hostname, commonly just the IP itself.
```

Comment:
Similarly to the above, but this time a wrong hostname was used to resolve to your IP address. This will lead to an invalid request to headscale.

Reply:
I don't think it's strictly true that a "hostname was used", since the log lines I observed actually listed my external IP as the "host".

This is why I chose the general phrasing "incorrect hostname" since I was hoping that could be interpreted as both an incorrect but actual hostname, or that the request uses "ip-as-hostname" which seemed to be the case for my logs.

Reply:
For my log most of the time it was the case with a different hostname. I guess both cases are valid.
```
2024/02/13 18:02:59 http: TLS handshake error from<IP_reacted>:45250: acme/autocert: host "<ip_reacted>" not configured in HostWhitelist
```
and
```
acme/autocert: host "<subdomain>.your-server.de" not configured in HostWhitelist
```